mne/preprocessing/_css.py

```diff
@@ -0,0 +1,123 @@
+# Author: John Samuelsson <johnsam@mit.edu>
+
+import numpy as np
+from ..utils import logger
+from ..evoked import EvokedArray, Evoked
+from .. import pick_types
+from numpy import linalg
+
+
+def _temp_proj(ref_2, ref_1, raw_data, n_proj=6):
+    """Remove common signal subspace of ref_2 and ref_1 from raw_data.
+
+    Parameters
+    ----------
+    ref_2 : np.ndarray of float, shape (n_sensors_2, n_times)
+        The magnetometer data for CSS. Can use either all magnetometer data or
+        a few selected sensors close to a region to be suppressed.
+    ref_1 : np.ndarray of float, shape (n_sensors_1, n_times)
+        The gradiometer data for CSS. Can use either all gradiometer data or
+        a few selected sensors close to a region to be suppressed.
+    raw_data : np.ndarray of float, shape (n_sensors_raw, n_times)
+        The data to be filtered, typically the EEG data.
+    n_proj : int
+        The number of projection vectors.
+
+    Returns
+    -------
+    filtered_data : np.ndarray of float, shape (n_sensors_raw, n_times)
+        The filtered data.
+
+    Notes
+    -----
+    This temporal projection procedure removes the common signal subspace
+    between ref_2 and ref_1 from raw_data using n_proj number of
+    projection vectors. Normally used for cortical signal suppression, where
+    ref_1 is gradiometer data, ref_2 is magnetometer data and
+    raw_data is EEG data.
+
+    """
+    # Orthonormalize gradiometer and magnetometer data by a QR decomposition
+    ref_1_orth = linalg.qr(ref_1.T)[0]
+    ref_2_orth = linalg.qr(ref_2.T)[0]
+
+    # Calculate cross-correlation
+    cross_corr = np.dot(ref_1_orth.T, ref_2_orth)
+
+    # Channel weights for common temporal subspace by SVD of cross-correlation
+    ref_1_ch_weights, sing_vals, ref_2_ch_weights = linalg.svd(cross_corr)
+
+    # Get temporal signals from channel weights
+    proj_mat = np.dot(ref_1_orth, ref_1_ch_weights)
+
+    # Project out common subspace
+    filtered_data = raw_data
+    for i in range(n_proj):
+        proj_vec = proj_mat[:, i].reshape(proj_mat.shape[0], 1)
+        weights = np.dot(filtered_data, proj_vec)
+        filtered_data = filtered_data - np.dot(weights, proj_vec.T)
+
+    return filtered_data
+
+
+def cortical_signal_suppression(evoked, mag_inds=None, grad_inds=None,
+                                n_proj=6):
+    """Remove the cortical signals in EEG by cortical signal suppression (CSS).
```

Comment:
EEG data is required? I thought this method could be used with just MEG data. It seems like by default it should use the mag+grad channels to remove cortical signals from all three channel types (mag, grad, eeg), in other words, always remove from mag+grad and also remove from eeg if present.